ufmt/core.py

```diff
@@ -2,118 +2,178 @@
 
 from .config import ufmt_config
+from .types import (
+    BlackConfig,
+    BlackConfigFactory,
+    Encoding,
+    FileContent,
+    Result,
+    UsortConfig,
+    UsortConfigFactory,
+)
+from .util import make_black_config, read_file, write_file
 
 LOG = logging.getLogger(__name__)
 
 
-@dataclass
-class Result:
-    path: Path
-    changed: bool = False
-    written: bool = False
-    diff: Optional[str] = None
-
-
-def ufmt_string(
+def ufmt_bytes(
     path: Path,
-    content: str,
+    content: FileContent,
+    *,
+    black_config: BlackConfig,
     usort_config: UsortConfig,
-    black_config: Optional[Mode] = None,
-) -> str:
-    content = usort_string(content, usort_config, path)
-    mode = black_config or Mode()
+    encoding: Encoding = "utf-8",
+) -> FileContent:
+    """
+    Format arbitrary bytes for the given path.
+
+    Requires passing valid config objects for both black and µsort. If the given path
+    represents a type stub (has a ``.pyi`` suffix), the black config object will be
+    updated to set ``is_pyi = True``.
+
+    Note: content will be decoded before passing to black, and re-encoded to bytes
+    again. Be sure to pass a known-valid unicode encoding for the content.
+    :func:`ufmt.util.read_file` can be used to both read bytes from disk, and make a
+    best guess at file encodings.
+    """
+    result = usort(content, usort_config, path)
 
     if path.suffix == ".pyi":
-        mode.is_pyi = True
+        black_config.is_pyi = True
 
     try:
-        content = format_file_contents(content, fast=False, mode=mode)
-    except NothingChanged:
-        pass
-
-    return content
-
-
-def make_black_config(path: Path) -> Mode:
-    config_file = find_pyproject_toml((str(path),))
-    if not config_file:
-        return Mode()
-
-    config = parse_pyproject_toml(config_file)
+        content_str = result.output.decode(encoding)
+        content_str = format_file_contents(content_str, fast=False, mode=black_config)
+        content = content_str.encode(encoding)
+        return content
 
-    # manually patch options that do not have a 1-to-1 match in Mode arguments
-    config["target_versions"] = {
-        TargetVersion[val.upper()] for val in config.pop("target_version", [])
-    }
-    config["string_normalization"] = not config.pop("skip_string_normalization", False)
-
-    names = {
-        field.name
-        for field in Mode.__dataclass_fields__.values()  # type: ignore[attr-defined]
-    }
-    config = {name: value for name, value in config.items() if name in names}
-
-    return Mode(**config)
+    except NothingChanged:
+        return result.output
 
 
-def ufmt_file(path: Path, dry_run: bool = False, diff: bool = False) -> Result:
-    usort_config = UsortConfig.find(path)
-    black_config = make_black_config(path)
+def ufmt_string(
+    path: Path,
+    content: str,
+    usort_config: UsortConfig,
+    black_config: Optional[BlackConfig] = None,
+    encoding: Encoding = "utf-8",
```

Comment:
I'd like a scarier comment about `encoding` here since it can't DTRT automatically.  The one at https://github.com/facebookexperimental/usort/blob/main/usort/api.py#L81 is a start, but now that I read the source it uses the system encoding, not utf-8.